typing/typecore.ml

```diff
@@ -2832,23 +2860,40 @@ let collect_apply_args env funct ignore_labels ty_fun ty_fun0 sargs =
                 if not optional && is_optional l' then
                   Location.prerr_warning sarg.pexp_loc
                     (Warnings.Nonoptional_label (Asttypes.string_of_label l));
-                remaining_sargs, use_arg sarg l'
+                remaining_sargs, Some (sarg, l')
             | None ->
-                sargs,
-                if optional && List.mem_assoc Nolabel sargs then
-                  eliminate_optional_arg ()
-                else begin
-                  (* No argument was given for this parameter, we abstract over
-                     it. *)
-                  may_warn funct.exp_loc
-                    (Warnings.Non_principal_labels "commuted an argument");
-                  Omitted { ty_arg; level = lv }
-                end
+                sargs, None
         in
-        loop ty_ret ty_ret0 ((l, arg) :: rev_args) remaining_sargs
-    | _ ->
-      (* We're not looking at a *known* function type anymore. *)
-      collect_unknown_apply_args env funct ty_fun0 rev_args sargs
+        match arrow_kind with
+        | `Arrow (ty_arg, ty_ret, ty_arg0, ty_ret0) ->
+            let use_arg sarg l' =
+              let wrapped_in_some = optional && not (is_optional l') in
+              if wrapped_in_some then
+                may_warn sarg.pexp_loc
+                  (not_principal "using an optional argument here");
+              Arg (Known_arg { sarg; ty_arg; ty_arg0; wrapped_in_some })
+            in
+            let eliminate_optional_arg () =
+              may_warn funct.exp_loc
+                (Warnings.Non_principal_labels "eliminated optional argument");
+              Arg (Eliminated_optional_arg { ty_arg; level = lv })
+            in
```

Comment:
In the current code, `eliminate_optional_arg` and `use_arg` are used once, and the names of the function does not seem that informative to me. Maybe they should be inlined?